src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1423,10 +1444,10 @@ ExpressionAnalysisResult::ExpressionAnalysisResult(
 
         if (query_analyzer.hasTableJoin())
         {
-            query_analyzer.appendJoinLeftKeys(chain, only_types || !first_stage);
-
+            Block left_block_sample = source_header;
```

Comment:
Note that `source_header` may already be not correct here, e.g. in case if query has `ARRAY JOIN`.
There should be used `ExpressionAnalyzerData::columns_after_array_join`.
But as I commented before, it's better to use `chain.getLastStep().getResultColumns()` inside `query_analyzer.appendJoinLeftKeys`

Reply:
`chain.getLastStep().getResultColumns` helped